src/main/java/org/broadinstitute/consent/http/service/DatasetService.java

```diff
@@ -100,7 +99,62 @@ public Dataset findDatasetByIdentifier(String datasetIdentifier) throws IllegalA
     if (!Objects.equals(d.getDatasetIdentifier(), datasetIdentifier)) {
       return null;
     }
-    return d;
+    return verifyPublicVisibilityAccess(d, user);
+  }
+
+  protected Dataset verifyPublicVisibilityAccess(Dataset dataset, User user) {
+    // Admins
+    if (user.hasUserRole(UserRoles.ADMIN)) {
+      return dataset;
+    }
+    // If there is no study, we can't verify visibility, so return the dataset
+    if (dataset.getStudyId() == null) {
+      return dataset;
+    }
+    // Study isn't always populated, so fetch it if necessary
+    if (dataset.getStudy() == null) {
+      dataset.setStudy(studyDAO.findStudyById(dataset.getStudyId()));
+    }
+    // If not visible, check that the user is authorized to see it
+    if (Boolean.FALSE.equals(dataset.getStudy().getPublicVisibility())) {
```

Comment:
@rushtong Can publicVisibility be mutated?

Reply:
If so, should we remove the study from Elastic if the value switches from ON -> OFF?

Reply:
> If so, should we remove the study from Elastic if the value switches from ON -> OFF?

Yes, study updates happen here: https://github.com/DataBiosphere/consent/blob/6c46cbeb42732a1b994e75ee50c00cab24e17594/src/main/java/org/broadinstitute/consent/http/resources/StudyResource.java#L210C19-L210C44

That code re-indexes the study here: https://github.com/DataBiosphere/consent/blob/6c46cbeb42732a1b994e75ee50c00cab24e17594/src/main/java/org/broadinstitute/consent/http/resources/StudyResource.java#L233